// teselagen-react-components/lib/MultiSelectSideBySide/items/item.js
"use strict";

exports.__esModule = true;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require("classnames");

var _classnames3 = _interopRequireDefault(_classnames2);

var _core = require("@blueprintjs/core");

var _item_label = require("./item_label");

var _item_label2 = _interopRequireDefault(_item_label);

require("./item.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Item = function Item(_ref) {
  var _classnames;

  var item = _ref.item,
      height = _ref.height,
      _onClick = _ref.onClick,
      withBorder = _ref.withBorder,
      checked = _ref.checked,
      indeterminate = _ref.indeterminate,
      disabled = _ref.disabled;
  return (
    /* eslint-disable no-useless-computed-key*/

    _react2.default.createElement(
      "div",
      {
        className: (0, _classnames3.default)("mss-item", (_classnames = {}, _classnames["mss-with_border"] = withBorder, _classnames["mss-selected"] = checked, _classnames["mss-disabled"] = disabled, _classnames)),
        style: { height: height },
        onClick: function onClick(e) {
          if (_onClick) {
            e.stopPropagation();
            e.preventDefault();
            _onClick(e);
          }
        }
        /* eslint-enable no-useless-computed-key*/
      },
      _react2.default.createElement(_core.Checkbox, {
        style: {
          marginTop: "10px",
          marginLeft: "10px"
        },
        type: "checkbox",
        color: "primary",
        checked: checked,
        indeterminate: indeterminate,
        disabled: disabled
      }),
      _react2.default.createElement(_item_label2.default, { label: item.label })
    )
  );
};

Item.propTypes = process.env.NODE_ENV !== "production" ? {
  item: _propTypes2.default.object,
  height: _propTypes2.default.number,
  withBorder: _propTypes2.default.bool,
  checked: _propTypes2.default.bool,
  indeterminate: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool
} : {};

Item.defaultProps = {
  item: {},
  height: 40,
  withBorder: false,
  checked: false,
  indeterminate: false,
  disabled: false
};

exports.default = Item;
module.exports = exports["default"];
