// teselagen-react-components/lib/MultiSelectSideBySide/items/select_all.js
"use strict";

exports.__esModule = true;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _item = require("./item");

var _item2 = _interopRequireDefault(_item);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SelectAll = function SelectAll(_ref) {
  var height = _ref.height,
      onClick = _ref.onClick,
      isAllSelected = _ref.isAllSelected,
      selectAllMessage = _ref.selectAllMessage,
      selectedIds = _ref.selectedIds;
  return _react2.default.createElement(_item2.default, {
    height: height,
    onClick: onClick,
    withBorder: true,
    item: { label: selectAllMessage },
    checked: isAllSelected,
    indeterminate: !isAllSelected && selectedIds.length > 0
  });
};

SelectAll.propTypes = process.env.NODE_ENV !== "production" ? {
  selectAllMessage: _propTypes2.default.string,
  height: _propTypes2.default.number,
  onClick: _propTypes2.default.func,
  isAllSelected: _propTypes2.default.bool,
  selectedIds: _propTypes2.default.arrayOf(_propTypes2.default.number)
} : {};

SelectAll.defaultProps = {
  selectAllMessage: "Select All",
  isAllSelected: false,
  height: 40,
  selectedIds: []
};

exports.default = SelectAll;
module.exports = exports["default"];
