// tg-client-query-builder/scratch/example-1.js

const QueryBuilder = require('../query-builder');

module.exports = function () {
    var qb = new QueryBuilder();

    var qry = qb.filter('Author')
        .whereAll({
            firstName: "Adam",
            lastName: "Thomas"
        }, {
            city: "Carmel"
        }, {
            id: qb.related("Post.Author_id")
                .whereAll({
                    likes: qb.greaterThan(10)
                }),

            Group_id: qb.related("Group.id")
                .whereAll({
                    groupType: qb.inList(['admin', 'supervisor'])
                })
        },
        {
            id: qb.related("AuthorRole.Author_id")
                .whereAll({
                    Role_id: qb.related("Role.id")
                        .whereAll({
                            name: 'Super-Author'
                        })
                })

        },
        {
            AuthorType_code: qb.notRelated("AuthorType.code")
                .whereAll({
                    name: 'Retired'
                }, {
                    name: 'Deceased'
                })
        })
        .andWhereAny({
            state: 'IN',
            country: 'USA'
        }, {
            zipCode: '46032'
        }, {
            age: qb.greaterThan(30)
        });

    var qryJson = qry.toJSON();
    return qryJson;
}
