// teselagen-react-components/es/MultiSelectSideBySide/destination_list.js
import React from "react";
import PropTypes from "prop-types";

import Column from "./column/column";
import List from "./list/items_list";
import NoItems from "./items/no_items";
import SelectedItem from "./items/selected_item";
import SelectionStatus from "./selection_status/selection_status";

var DestinationList = function DestinationList(_ref) {
  var selectionStatusRenderer = _ref.selectionStatusRenderer,
      selectedIds = _ref.selectedIds,
      clearAll = _ref.clearAll,
      messages = _ref.messages,
      selectedItems = _ref.selectedItems,
      itemHeight = _ref.itemHeight,
      height = _ref.height,
      unselectItems = _ref.unselectItems,
      selectedItemRenderer = _ref.selectedItemRenderer,
      noItemsRenderer = _ref.noItemsRenderer;

  var SelectionStatusRenderer = selectionStatusRenderer;
  return React.createElement(
    Column,
    null,
    React.createElement(SelectionStatusRenderer, {
      selected: selectedIds,
      clearAll: clearAll,
      clearAllMessage: messages.clearAllMessage,
      selectedMessage: messages.selectedMessage,
      noneSelectedMessage: messages.noneSelectedMessage
    }),
    React.createElement(List, {
      items: selectedItems,
      itemHeight: itemHeight,
      height: height - 45,
      onClick: function onClick(event, id) {
        return unselectItems([id]);
      },
      renderer: selectedItemRenderer,
      noItemsRenderer: noItemsRenderer,
      noItemsMessage: messages.noItemsMessage
    })
  );
};

DestinationList.propTypes = process.env.NODE_ENV !== "production" ? {
  selectionStatusRenderer: PropTypes.any,
  selectedIds: PropTypes.arrayOf(PropTypes.number),
  clearAll: PropTypes.func,
  messages: PropTypes.object,
  selectedItems: PropTypes.array,
  itemHeight: PropTypes.number,
  height: PropTypes.number,
  unselectItems: PropTypes.func,
  selectedItemRenderer: PropTypes.any,
  noItemsRenderer: PropTypes.any
} : {};

DestinationList.defaultProps = {
  selectionStatusRenderer: SelectionStatus,
  selectedIds: [],
  messages: {},
  selectedItems: [],
  itemHeight: 40,
  height: 400,
  selectedItemRenderer: SelectedItem,
  noItemsRenderer: NoItems
};

export default DestinationList;
