// teselagen-react-components/es/MultiSelectSideBySide/index.js
var _class, _temp;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from "react";
import classnames from "classnames";
import PropTypes from "prop-types";

import withMultiSelectState from "./multi_select_state";
import SourceList from "./source_list";
import DestinationList from "./destination_list";

import "./multi_select.css";
import Loader from "./loader/loader";

export var MultiSelect = (_temp = _class = function (_PureComponent) {
  _inherits(MultiSelect, _PureComponent);

  function MultiSelect() {
    _classCallCheck(this, MultiSelect);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  MultiSelect.prototype.calculateHeight = function calculateHeight() {
    var _props = this.props,
        height = _props.height,
        showSearch = _props.showSearch,
        showSelectAll = _props.showSelectAll,
        itemHeight = _props.itemHeight,
        selectAllHeight = _props.selectAllHeight,
        maxSelectedItems = _props.maxSelectedItems;

    height = showSearch && !maxSelectedItems ? height - 45 : height;
    height = showSelectAll ? height - (selectAllHeight ? selectAllHeight : itemHeight) : height;
    return height;
  };

  MultiSelect.prototype.render = function render() {
    var _props2 = this.props,
        wrapperClassName = _props2.wrapperClassName,
        showSearch = _props2.showSearch,
        height = _props2.height,
        itemHeight = _props2.itemHeight,
        selectAllHeight = _props2.selectAllHeight,
        showSelectAll = _props2.showSelectAll,
        showSelectedItems = _props2.showSelectedItems,
        itemRenderer = _props2.itemRenderer,
        selectedItemRenderer = _props2.selectedItemRenderer,
        filteredItems = _props2.filteredItems,
        selectedItems = _props2.selectedItems,
        getList = _props2.getList,
        filterItems = _props2.filterItems,
        selectAllItems = _props2.selectAllItems,
        isAllSelected = _props2.isAllSelected,
        clearAll = _props2.clearAll,
        selectItem = _props2.selectItem,
        unselectItems = _props2.unselectItems,
        searchIcon = _props2.searchIcon,
        searchRenderer = _props2.searchRenderer,
        selectAllRenderer = _props2.selectAllRenderer,
        selectionStatusRenderer = _props2.selectionStatusRenderer,
        noItemsRenderer = _props2.noItemsRenderer,
        loaderRenderer = _props2.loaderRenderer,
        messages = _props2.messages,
        loading = _props2.loading,
        maxSelectedItems = _props2.maxSelectedItems,
        searchValue = _props2.searchValue;

    var calculatedHeight = this.calculateHeight();
    var selectedIds = selectedItems.map(function (item) {
      return item.id;
    });
    var disabled = maxSelectedItems && maxSelectedItems <= selectedItems.length;
    var LoaderRenderer = loaderRenderer;
    return React.createElement(
      "div",
      {
        className: classnames("mss-wrapper", wrapperClassName),
        style: { height: height }
      },
      loading && React.createElement(LoaderRenderer, null),
      !loading && React.createElement(SourceList, {
        searchValue: searchValue,
        searchRenderer: searchRenderer,
        selectAllRenderer: selectAllRenderer,
        showSearch: showSearch,
        filterItems: filterItems,
        searchIcon: searchIcon,
        messages: messages,
        showSelectAll: showSelectAll && !maxSelectedItems,
        itemHeight: itemHeight,
        selectAllItems: selectAllItems,
        isAllSelected: isAllSelected,
        selectedIds: selectedIds,
        itemRenderer: itemRenderer,
        getList: getList,
        filteredItems: filteredItems,
        calculatedHeight: calculatedHeight,
        selectItem: selectItem,
        noItemsRenderer: noItemsRenderer,
        disabled: disabled,
        selectAllHeight: selectAllHeight
      }),
      !loading && showSelectedItems && React.createElement(DestinationList, {
        selectionStatusRenderer: selectionStatusRenderer,
        selectedIds: selectedIds,
        clearAll: clearAll,
        messages: messages,
        selectedItems: selectedItems,
        itemHeight: itemHeight,
        height: height,
        unselectItems: unselectItems,
        selectedItemRenderer: selectedItemRenderer,
        noItemsRenderer: noItemsRenderer
      })
    );
  };

  return MultiSelect;
}(PureComponent), _class.defaultProps = {
  selectedItems: [],
  filteredItems: [],
  messages: {},
  showSearch: true,
  showSelectAll: true,
  showSelectedItems: true,
  height: 400,
  itemHeight: 40,
  loaderRenderer: Loader
}, _temp);

MultiSelect.propTypes = process.env.NODE_ENV !== "production" ? {
  selectedItems: PropTypes.array,
  filteredItems: PropTypes.array,
  loading: PropTypes.bool,
  messages: PropTypes.object,
  onChange: PropTypes.func,
  showSearch: PropTypes.bool,
  showSelectAll: PropTypes.bool,
  showSelectedItems: PropTypes.bool,
  searchIcon: PropTypes.string,
  deleteIcon: PropTypes.string,
  searchRenderer: PropTypes.func,
  selectedItemRenderer: PropTypes.any,
  height: PropTypes.number,
  itemHeight: PropTypes.number,
  selectAllHeight: PropTypes.number,
  loaderRenderer: PropTypes.any,
  maxSelectedItems: PropTypes.number
} : {};
export default withMultiSelectState(MultiSelect);
