// teselagen-react-components/es/MultiSelectSideBySide/search/search.js
import React from "react";

import PropTypes from "prop-types";
import { Icon } from "@blueprintjs/core";

import "./search.css";

var Search = function Search(_ref) {
  var searchPlaceholder = _ref.searchPlaceholder,
      searchIcon = _ref.searchIcon,
      onChange = _ref.onChange;

  return React.createElement(
    "div",
    { className: "mss-search" },
    React.createElement("input", {
      type: "text",
      className: "mss-input",
      placeholder: searchPlaceholder,
      onChange: onChange
    }),
    React.createElement(
      "div",
      { className: "mss-icon" },
      React.createElement(Icon, { icon: searchIcon })
    )
  );
};

export var SearchWithValue = function SearchWithValue(_ref2) {
  var searchPlaceholder = _ref2.searchPlaceholder,
      onChange = _ref2.onChange,
      value = _ref2.value,
      searchIcon = _ref2.searchIcon;

  return React.createElement(
    "div",
    { className: "mss-search" },
    React.createElement("input", {
      value: value,
      type: "text",
      className: "mss-input",
      placeholder: searchPlaceholder,
      onChange: onChange
    }),
    React.createElement(
      "div",
      { className: "mss-icon" },
      React.createElement(Icon, { icon: searchIcon })
    )
  );
};

Search.propTypes = process.env.NODE_ENV !== "production" ? {
  searchPlaceholder: PropTypes.string,
  searchIcon: PropTypes.string,
  onChange: PropTypes.func
} : {};

Search.defaultProps = {
  searchPlaceholder: "Search...",
  searchIcon: "search"
};

export default Search;
