// teselagen-react-components/lib/MultiSelectSideBySide/destination_list.js
"use strict";

exports.__esModule = true;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _column = require("./column/column");

var _column2 = _interopRequireDefault(_column);

var _items_list = require("./list/items_list");

var _items_list2 = _interopRequireDefault(_items_list);

var _no_items = require("./items/no_items");

var _no_items2 = _interopRequireDefault(_no_items);

var _selected_item = require("./items/selected_item");

var _selected_item2 = _interopRequireDefault(_selected_item);

var _selection_status = require("./selection_status/selection_status");

var _selection_status2 = _interopRequireDefault(_selection_status);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DestinationList = function DestinationList(_ref) {
  var selectionStatusRenderer = _ref.selectionStatusRenderer,
      selectedIds = _ref.selectedIds,
      clearAll = _ref.clearAll,
      messages = _ref.messages,
      selectedItems = _ref.selectedItems,
      itemHeight = _ref.itemHeight,
      height = _ref.height,
      unselectItems = _ref.unselectItems,
      selectedItemRenderer = _ref.selectedItemRenderer,
      noItemsRenderer = _ref.noItemsRenderer;

  var SelectionStatusRenderer = selectionStatusRenderer;
  return _react2.default.createElement(
    _column2.default,
    null,
    _react2.default.createElement(SelectionStatusRenderer, {
      selected: selectedIds,
      clearAll: clearAll,
      clearAllMessage: messages.clearAllMessage,
      selectedMessage: messages.selectedMessage,
      noneSelectedMessage: messages.noneSelectedMessage
    }),
    _react2.default.createElement(_items_list2.default, {
      items: selectedItems,
      itemHeight: itemHeight,
      height: height - 45,
      onClick: function onClick(event, id) {
        return unselectItems([id]);
      },
      renderer: selectedItemRenderer,
      noItemsRenderer: noItemsRenderer,
      noItemsMessage: messages.noItemsMessage
    })
  );
};

DestinationList.propTypes = process.env.NODE_ENV !== "production" ? {
  selectionStatusRenderer: _propTypes2.default.any,
  selectedIds: _propTypes2.default.arrayOf(_propTypes2.default.number),
  clearAll: _propTypes2.default.func,
  messages: _propTypes2.default.object,
  selectedItems: _propTypes2.default.array,
  itemHeight: _propTypes2.default.number,
  height: _propTypes2.default.number,
  unselectItems: _propTypes2.default.func,
  selectedItemRenderer: _propTypes2.default.any,
  noItemsRenderer: _propTypes2.default.any
} : {};

DestinationList.defaultProps = {
  selectionStatusRenderer: _selection_status2.default,
  selectedIds: [],
  messages: {},
  selectedItems: [],
  itemHeight: 40,
  height: 400,
  selectedItemRenderer: _selected_item2.default,
  noItemsRenderer: _no_items2.default
};

exports.default = DestinationList;
module.exports = exports["default"];
