// teselagen-react-components/lib/MultiSelectSideBySide/source_list.js
"use strict";

exports.__esModule = true;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _column = require("./column/column");

var _column2 = _interopRequireDefault(_column);

var _items_list = require("./list/items_list");

var _items_list2 = _interopRequireDefault(_items_list);

var _no_items = require("./items/no_items");

var _no_items2 = _interopRequireDefault(_no_items);

var _search = require("./search/search");

var _search2 = _interopRequireDefault(_search);

var _select_all = require("./items/select_all");

var _select_all2 = _interopRequireDefault(_select_all);

var _item = require("./items/item");

var _item2 = _interopRequireDefault(_item);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SourceList = function SourceList(_ref) {
  var searchRenderer = _ref.searchRenderer,
      selectAllRenderer = _ref.selectAllRenderer,
      showSearch = _ref.showSearch,
      filterItems = _ref.filterItems,
      searchIcon = _ref.searchIcon,
      messages = _ref.messages,
      showSelectAll = _ref.showSelectAll,
      itemHeight = _ref.itemHeight,
      selectAllHeight = _ref.selectAllHeight,
      selectAllItems = _ref.selectAllItems,
      isAllSelected = _ref.isAllSelected,
      selectedIds = _ref.selectedIds,
      itemRenderer = _ref.itemRenderer,
      getList = _ref.getList,
      filteredItems = _ref.filteredItems,
      calculatedHeight = _ref.calculatedHeight,
      selectItem = _ref.selectItem,
      noItemsRenderer = _ref.noItemsRenderer,
      disabled = _ref.disabled,
      searchValue = _ref.searchValue;

  var SearchRenderer = searchRenderer;
  var SelectAllRenderer = selectAllRenderer;
  return _react2.default.createElement(
    _column2.default,
    null,
    showSearch && _react2.default.createElement(SearchRenderer, {
      onChange: filterItems,
      searchIcon: searchIcon,
      value: searchValue,
      searchPlaceholder: messages.searchPlaceholder
    }),
    showSelectAll && _react2.default.createElement(SelectAllRenderer, {
      height: selectAllHeight ? selectAllHeight : itemHeight,
      onClick: selectAllItems,
      isAllSelected: isAllSelected,
      selectedIds: selectedIds,
      renderer: itemRenderer,
      selectAllMessage: messages.selectAllMessage
    }),
    _react2.default.createElement(_items_list2.default, {
      ref: getList,
      offset: 1,
      items: filteredItems,
      itemHeight: itemHeight,
      height: calculatedHeight,
      onClick: selectItem,
      selectedIds: selectedIds,
      renderer: itemRenderer,
      noItemsRenderer: noItemsRenderer,
      noItemsMessage: messages.noItemsMessage,
      disabled: disabled,
      disabledItemsTooltip: messages.disabledItemsTooltip
    })
  );
};

SourceList.propTypes = process.env.NODE_ENV !== "production" ? {
  searchRenderer: _propTypes2.default.any,
  selectAllRenderer: _propTypes2.default.any,
  noItemsRenderer: _propTypes2.default.any,
  itemRenderer: _propTypes2.default.any,
  searchIcon: _propTypes2.default.any,
  showSearch: _propTypes2.default.bool,
  showSelectAll: _propTypes2.default.bool,
  isAllSelected: _propTypes2.default.bool,
  filterItems: _propTypes2.default.func,
  messages: _propTypes2.default.object,
  itemHeight: _propTypes2.default.number,
  selectAllHeight: _propTypes2.default.number,
  calculatedHeight: _propTypes2.default.number,
  filteredItems: _propTypes2.default.array,
  selectedIds: _propTypes2.default.arrayOf(_propTypes2.default.number),
  selectAllItems: _propTypes2.default.func,
  getList: _propTypes2.default.func,
  selectItem: _propTypes2.default.func,
  disabled: _propTypes2.default.bool
} : {};

SourceList.defaultProps = {
  searchRenderer: _search2.default,
  selectAllRenderer: _select_all2.default,
  noItemsRenderer: _no_items2.default,
  itemRenderer: _item2.default,
  showSearch: true,
  showSelectAll: true,
  isAllSelected: false,
  calculatedHeight: 400,
  itemHeight: 40,
  selectedIds: [],
  filteredItems: [],
  messages: {},
  disabled: false
};

exports.default = SourceList;
module.exports = exports["default"];
