// teselagen-react-components/lib/DataTable/defaultProps.js
"use strict";

exports.__esModule = true;

var _lodash = require("lodash");

exports.default = {
  //NOTE: DO NOT SET DEFAULTS HERE FOR PROPS THAT GET COMPUTED AS PART OF PRESET GROUPS IN computePresets
  addFilters: _lodash.noop,
  className: "",
  clearFilters: _lodash.noop,
  contextMenu: _lodash.noop,
  disabled: false,
  entities: [],
  extraClasses: "",
  filters: [],
  isCopyable: true,
  isEntityDisabled: _lodash.noop,
  isLoading: false,
  isSimple: false,
  isSingleSelect: false,
  maxHeight: 600,
  noHeader: false,
  noSelect: false,
  noUserSelect: false,
  onDeselect: _lodash.noop,
  onDoubleClick: _lodash.noop,
  onMultiRowSelect: _lodash.noop,
  onRowClick: _lodash.noop,
  onRowSelect: _lodash.noop,
  onSingleRowSelect: _lodash.noop,
  page: 1,
  pageSize: 10,
  reduxFormExpandedEntityIdMap: {},
  reduxFormSearchInput: "",
  reduxFormSelectedEntityIdMap: {},
  removeSingleFilter: _lodash.noop,
  resized: [],
  resizePersist: _lodash.noop,
  setFilter: _lodash.noop,
  setOrder: _lodash.noop,
  setPage: _lodash.noop,
  setPageSize: _lodash.noop,
  setSearchTerm: _lodash.noop,
  showCount: false,
  style: {},
  withCheckboxes: false,
  withSort: true
};
module.exports = exports["default"];
