// teselagen-react-components/lib/DataTable/SearchBar.js
"use strict";

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _core = require("@blueprintjs/core");

var _classnames = require("classnames");

var _classnames2 = _interopRequireDefault(_classnames);

var _handlerHelpers = require("../utils/handlerHelpers");

var _FormComponents = require("../FormComponents");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SearchBar = function SearchBar(_ref) {
  var reduxFormSearchInput = _ref.reduxFormSearchInput,
      setSearchTerm = _ref.setSearchTerm,
      loading = _ref.loading,
      searchMenuButton = _ref.searchMenuButton,
      disabled = _ref.disabled;

  return _react2.default.createElement(_FormComponents.InputField, _extends({
    autoFocus: true,
    disabled: disabled,
    loading: loading,
    type: "search",
    name: "reduxFormSearchInput",
    className: (0, _classnames2.default)("datatable-search-input", _core.Classes.ROUND),
    placeholder: "Search..."
  }, (0, _handlerHelpers.onEnterHelper)(function (e) {
    e.preventDefault();
    setSearchTerm(reduxFormSearchInput);
  }), {
    rightElement: loading ? _react2.default.createElement(_core.Spinner, { size: "18" }) : searchMenuButton || _react2.default.createElement(_core.Button, {
      minimal: true,
      icon: "search",
      onClick: function onClick() {
        setSearchTerm(reduxFormSearchInput);
      }
    })
  }));
};

exports.default = SearchBar;
module.exports = exports["default"];
