// teselagen-react-components/es/MultiSelectSideBySide/selection_status/selection_status.js
import React from "react";
import PropTypes from "prop-types";

import "./selection_status.css";

var SelectionStatus = function SelectionStatus(_ref) {
  var selected = _ref.selected,
      clearAll = _ref.clearAll,
      clearAllMessage = _ref.clearAllMessage,
      noneSelectedMessage = _ref.noneSelectedMessage,
      selectedMessage = _ref.selectedMessage;
  return React.createElement(
    "div",
    { className: "mss-selection_status" },
    React.createElement(
      "div",
      { className: "mss-status" },
      selected.length > 0 ? selected.length + " " + selectedMessage : noneSelectedMessage
    ),
    React.createElement(
      "div",
      { className: "mss-clear_all", onClick: clearAll },
      selected.length > 0 ? clearAllMessage : ""
    )
  );
};

SelectionStatus.propTypes = process.env.NODE_ENV !== "production" ? {
  selected: PropTypes.array,
  clearAll: PropTypes.func,
  clearAllMessage: PropTypes.string,
  noneSelectedMessage: PropTypes.string,
  selectedMessage: PropTypes.string
} : {};

SelectionStatus.defaultProps = {
  selected: [],
  clearAllMessage: "Clear All",
  noneSelectedMessage: "None Selected",
  selectedMessage: "selected"
};

export default SelectionStatus;
