// teselagen-react-components/lib/ScrollToTop/index.js
"use strict";

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _core = require("@blueprintjs/core");

var _reactTransitionGroup = require("react-transition-group");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var duration = 300;

var defaultStyle = {
  transition: "opacity " + duration + "ms ease-in-out",
  opacity: 0
};

var transitionStyles = {
  entering: { opacity: 1 },
  entered: { opacity: 1 },
  exiting: { opacity: 0 },
  exited: { opacity: 0 }
};

function ScrollToTop(_ref) {
  var _ref$showAt = _ref.showAt,
      showAt = _ref$showAt === undefined ? 150 : _ref$showAt,
      _ref$scrollContainer = _ref.scrollContainer,
      scrollContainer = _ref$scrollContainer === undefined ? document.scrollingElement : _ref$scrollContainer;

  var _useState = (0, _react.useState)(scrollContainer.scrollTop >= showAt),
      shouldShow = _useState[0],
      setShouldShow = _useState[1];

  (0, _react.useEffect)(function () {
    var scrollListener = function scrollListener() {
      var newShouldShow = scrollContainer.scrollTop >= showAt;
      if (newShouldShow !== shouldShow) setShouldShow(newShouldShow);
    };
    var listenerContainer = scrollContainer === document.scrollingElement ? window : scrollContainer;
    listenerContainer.addEventListener("scroll", scrollListener);
    // Specify how to clean up after this effect:
    return function cleanup() {
      listenerContainer.removeEventListener("scroll", scrollListener);
    };
  }, [shouldShow, showAt, scrollContainer]);

  var scrollToTop = function scrollToTop() {
    var c = scrollContainer.scrollTop;
    if (c > 0) {
      window.requestAnimationFrame(scrollToTop);
      scrollContainer.scrollTo(0, c - c / 8);
    }
  };

  return _react2.default.createElement(
    _reactTransitionGroup.Transition,
    { "in": shouldShow, timeout: duration },
    function (state) {
      return state === "exited" ? null : _react2.default.createElement(
        "div",
        { style: { position: "fixed", bottom: 25, right: 25, zIndex: 10 } },
        _react2.default.createElement(_core.Button, {
          style: _extends({
            borderRadius: "50%"
          }, defaultStyle, transitionStyles[state]),
          intent: "primary",
          icon: "arrow-up",
          large: true,
          onClick: scrollToTop
        })
      );
    }
  );
}

exports.default = ScrollToTop;
module.exports = exports["default"];
