// teselagen-react-components/lib/MultiSelectSideBySide/list/items_list.js
"use strict";

exports.__esModule = true;

var _class, _temp;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _AutoSizer = require("react-virtualized/dist/commonjs/AutoSizer");

var _list = require("./list");

var _list2 = _interopRequireDefault(_list);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ItemsList = (_temp = _class = function (_PureComponent) {
  _inherits(ItemsList, _PureComponent);

  function ItemsList(props) {
    _classCallCheck(this, ItemsList);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props));

    _this.update = _this.update.bind(_this);
    _this.getlistRef = _this.getlistRef.bind(_this);
    return _this;
  }

  ItemsList.prototype.update = function update() {
    this.listRef.forceUpdateGrid();
  };

  ItemsList.prototype.componentDidUpdate = function componentDidUpdate() {
    this.update();
  };

  ItemsList.prototype.getlistRef = function getlistRef(ref) {
    this.listRef = ref;
  };

  ItemsList.prototype.render = function render() {
    var _this2 = this;

    var _props = this.props,
        height = _props.height,
        itemHeight = _props.itemHeight,
        items = _props.items,
        offset = _props.offset,
        noItemsMessage = _props.noItemsMessage,
        noItemsRenderer = _props.noItemsRenderer,
        renderer = _props.renderer,
        selectedIds = _props.selectedIds,
        onClick = _props.onClick,
        disabled = _props.disabled,
        disabledItemsTooltip = _props.disabledItemsTooltip;

    return _react2.default.createElement(
      _AutoSizer.AutoSizer,
      null,
      function (_ref) {
        var width = _ref.width;
        return _react2.default.createElement(_list2.default, {
          getlistRef: _this2.getlistRef,
          rowRenderer: _this2.rowRenderer,
          noRowsRenderer: _this2.noRowsRenderer,
          width: width - offset,
          rowHeight: itemHeight,
          height: height,
          rowCount: items.length,
          noItemsMessage: noItemsMessage,
          noItemsRenderer: noItemsRenderer,
          renderer: renderer,
          itemHeight: itemHeight,
          onClick: onClick,
          items: items,
          selectedIds: selectedIds,
          disabled: disabled,
          disabledItemsTooltip: disabledItemsTooltip
        });
      }
    );
  };

  return ItemsList;
}(_react.PureComponent), _class.defaultProps = {
  itemHeight: 40,
  height: 400,
  offset: 0,
  selectedIds: [],
  items: [],
  disabled: false
}, _temp);
ItemsList.propTypes = process.env.NODE_ENV !== "production" ? {
  renderer: _propTypes2.default.any,
  noItemsRenderer: _propTypes2.default.any,
  itemHeight: _propTypes2.default.number,
  height: _propTypes2.default.number,
  offset: _propTypes2.default.number,
  onClick: _propTypes2.default.func,
  selectedIds: _propTypes2.default.arrayOf(_propTypes2.default.number),
  items: _propTypes2.default.array,
  disabled: _propTypes2.default.bool,
  disabledItemsTooltip: _propTypes2.default.string
} : {};
exports.default = ItemsList;
module.exports = exports["default"];
