// teselagen-react-components/lib/MultiSelectSideBySide/list/list.js
"use strict";

exports.__esModule = true;

var _class, _temp;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _List = require("react-virtualized/dist/commonjs/List");

require("./list.css");

var _item = require("../items/item");

var _item2 = _interopRequireDefault(_item);

var _no_items = require("../items/no_items");

var _no_items2 = _interopRequireDefault(_no_items);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var InnerList = (_temp = _class = function (_PureComponent) {
  _inherits(InnerList, _PureComponent);

  function InnerList(props) {
    _classCallCheck(this, InnerList);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props));

    _this.rowRenderer = _this.rowRenderer.bind(_this);
    _this.noRowsRenderer = _this.noRowsRenderer.bind(_this);
    _this.onClick = _this.onClick.bind(_this);
    return _this;
  }

  InnerList.prototype.onClick = function onClick(event, id, disabled) {
    var _props = this.props,
        onClick = _props.onClick,
        selectedIds = _props.selectedIds;

    var checked = selectedIds.includes(id);
    if (disabled && checked || !disabled) {
      onClick(event, id);
    }
  };

  InnerList.prototype.rowRenderer = function rowRenderer(_ref) {
    var _this2 = this;

    var index = _ref.index,
        key = _ref.key,
        style = _ref.style;
    var _props2 = this.props,
        renderer = _props2.renderer,
        itemHeight = _props2.itemHeight,
        items = _props2.items,
        selectedIds = _props2.selectedIds,
        disabledItemsTooltip = _props2.disabledItemsTooltip;

    var Renderer = renderer;
    var item = items[index];
    var checked = selectedIds.includes(item.id);
    var disabled = this.props.disabled || item.disabled;
    return _react2.default.createElement(
      "div",
      {
        key: key,
        style: style,
        className: "mss-list_item",
        onClick: function onClick(event) {
          return _this2.onClick(event, item.id, disabled);
        },
        title: disabled ? disabledItemsTooltip : undefined
      },
      _react2.default.createElement(Renderer, {
        item: item,
        onClick: function onClick(event) {
          return _this2.onClick(event, item.id, disabled);
        },
        height: itemHeight,
        checked: checked,
        disabled: disabled && !checked
      })
    );
  };

  InnerList.prototype.noRowsRenderer = function noRowsRenderer() {
    var _props3 = this.props,
        noItemsMessage = _props3.noItemsMessage,
        noItemsRenderer = _props3.noItemsRenderer;

    var NoItemsRenderer = noItemsRenderer;
    return _react2.default.createElement(NoItemsRenderer, { noItemsMessage: noItemsMessage });
  };

  InnerList.prototype.render = function render() {
    var _props4 = this.props,
        height = _props4.height,
        itemHeight = _props4.itemHeight,
        items = _props4.items,
        offset = _props4.offset,
        width = _props4.width,
        getlistRef = _props4.getlistRef;

    return _react2.default.createElement(_List.List, {
      ref: getlistRef,
      className: "mss-list",
      rowRenderer: this.rowRenderer,
      noRowsRenderer: this.noRowsRenderer,
      width: width - offset,
      rowHeight: itemHeight,
      height: height,
      rowCount: items.length
    });
  };

  return InnerList;
}(_react.PureComponent), _class.defaultProps = {
  renderer: _item2.default,
  noItemsRenderer: _no_items2.default,
  itemHeight: 40,
  height: 400,
  offset: 0,
  selectedIds: [],
  items: [],
  disabled: false
}, _temp);
InnerList.propTypes = process.env.NODE_ENV !== "production" ? {
  renderer: _propTypes2.default.any,
  noItemsRenderer: _propTypes2.default.any,
  itemHeight: _propTypes2.default.number,
  height: _propTypes2.default.number,
  offset: _propTypes2.default.number,
  onClick: _propTypes2.default.func,
  selectedIds: _propTypes2.default.arrayOf(_propTypes2.default.number),
  items: _propTypes2.default.array,
  disabled: _propTypes2.default.bool,
  disabledItemsTooltip: _propTypes2.default.string
} : {};
exports.default = InnerList;
module.exports = exports["default"];
