// teselagen-react-components/lib/MultiSelectSideBySide/selection_status/selection_status.js
"use strict";

exports.__esModule = true;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

require("./selection_status.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SelectionStatus = function SelectionStatus(_ref) {
  var selected = _ref.selected,
      clearAll = _ref.clearAll,
      clearAllMessage = _ref.clearAllMessage,
      noneSelectedMessage = _ref.noneSelectedMessage,
      selectedMessage = _ref.selectedMessage;
  return _react2.default.createElement(
    "div",
    { className: "mss-selection_status" },
    _react2.default.createElement(
      "div",
      { className: "mss-status" },
      selected.length > 0 ? selected.length + " " + selectedMessage : noneSelectedMessage
    ),
    _react2.default.createElement(
      "div",
      { className: "mss-clear_all", onClick: clearAll },
      selected.length > 0 ? clearAllMessage : ""
    )
  );
};

SelectionStatus.propTypes = process.env.NODE_ENV !== "production" ? {
  selected: _propTypes2.default.array,
  clearAll: _propTypes2.default.func,
  clearAllMessage: _propTypes2.default.string,
  noneSelectedMessage: _propTypes2.default.string,
  selectedMessage: _propTypes2.default.string
} : {};

SelectionStatus.defaultProps = {
  selected: [],
  clearAllMessage: "Clear All",
  noneSelectedMessage: "None Selected",
  selectedMessage: "selected"
};

exports.default = SelectionStatus;
module.exports = exports["default"];
