// teselagen-react-components/es/MultiSelectSideBySide/list/list.js
var _class, _temp;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { List } from "react-virtualized/dist/commonjs/List";

import "./list.css";
import Item from "../items/item";
import NoItems from "../items/no_items";

var InnerList = (_temp = _class = function (_PureComponent) {
  _inherits(InnerList, _PureComponent);

  function InnerList(props) {
    _classCallCheck(this, InnerList);

    var _this = _possibleConstructorReturn(this, _PureComponent.call(this, props));

    _this.rowRenderer = _this.rowRenderer.bind(_this);
    _this.noRowsRenderer = _this.noRowsRenderer.bind(_this);
    _this.onClick = _this.onClick.bind(_this);
    return _this;
  }

  InnerList.prototype.onClick = function onClick(event, id, disabled) {
    var _props = this.props,
        onClick = _props.onClick,
        selectedIds = _props.selectedIds;

    var checked = selectedIds.includes(id);
    if (disabled && checked || !disabled) {
      onClick(event, id);
    }
  };

  InnerList.prototype.rowRenderer = function rowRenderer(_ref) {
    var _this2 = this;

    var index = _ref.index,
        key = _ref.key,
        style = _ref.style;
    var _props2 = this.props,
        renderer = _props2.renderer,
        itemHeight = _props2.itemHeight,
        items = _props2.items,
        selectedIds = _props2.selectedIds,
        disabledItemsTooltip = _props2.disabledItemsTooltip;

    var Renderer = renderer;
    var item = items[index];
    var checked = selectedIds.includes(item.id);
    var disabled = this.props.disabled || item.disabled;
    return React.createElement(
      "div",
      {
        key: key,
        style: style,
        className: "mss-list_item",
        onClick: function onClick(event) {
          return _this2.onClick(event, item.id, disabled);
        },
        title: disabled ? disabledItemsTooltip : undefined
      },
      React.createElement(Renderer, {
        item: item,
        onClick: function onClick(event) {
          return _this2.onClick(event, item.id, disabled);
        },
        height: itemHeight,
        checked: checked,
        disabled: disabled && !checked
      })
    );
  };

  InnerList.prototype.noRowsRenderer = function noRowsRenderer() {
    var _props3 = this.props,
        noItemsMessage = _props3.noItemsMessage,
        noItemsRenderer = _props3.noItemsRenderer;

    var NoItemsRenderer = noItemsRenderer;
    return React.createElement(NoItemsRenderer, { noItemsMessage: noItemsMessage });
  };

  InnerList.prototype.render = function render() {
    var _props4 = this.props,
        height = _props4.height,
        itemHeight = _props4.itemHeight,
        items = _props4.items,
        offset = _props4.offset,
        width = _props4.width,
        getlistRef = _props4.getlistRef;

    return React.createElement(List, {
      ref: getlistRef,
      className: "mss-list",
      rowRenderer: this.rowRenderer,
      noRowsRenderer: this.noRowsRenderer,
      width: width - offset,
      rowHeight: itemHeight,
      height: height,
      rowCount: items.length
    });
  };

  return InnerList;
}(PureComponent), _class.defaultProps = {
  renderer: Item,
  noItemsRenderer: NoItems,
  itemHeight: 40,
  height: 400,
  offset: 0,
  selectedIds: [],
  items: [],
  disabled: false
}, _temp);
InnerList.propTypes = process.env.NODE_ENV !== "production" ? {
  renderer: PropTypes.any,
  noItemsRenderer: PropTypes.any,
  itemHeight: PropTypes.number,
  height: PropTypes.number,
  offset: PropTypes.number,
  onClick: PropTypes.func,
  selectedIds: PropTypes.arrayOf(PropTypes.number),
  items: PropTypes.array,
  disabled: PropTypes.bool,
  disabledItemsTooltip: PropTypes.string
} : {};


export default InnerList;
