// @teselagen/apollo-methods/lib/index.js
"use strict";

exports.__esModule = true;

var _withDelete = require("./withDelete");

Object.defineProperty(exports, "withDelete", {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_withDelete).default;
  }
});

var _withUpsert = require("./withUpsert");

Object.defineProperty(exports, "withUpsert", {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_withUpsert).default;
  }
});

var _withQuery = require("./withQuery");

Object.defineProperty(exports, "withQuery", {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_withQuery).default;
  }
});

var _generateQuery = require("./utils/generateQuery");

Object.defineProperty(exports, "generateQuery", {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_generateQuery).default;
  }
});

var _generateFragmentWithFields = require("./utils/generateFragmentWithFields");

Object.defineProperty(exports, "generateFragmentWithFields", {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_generateFragmentWithFields).default;
  }
});

var _getApolloMethods = require("./getApolloMethods");

Object.defineProperty(exports, "getApolloMethods", {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_getApolloMethods).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
