// teselagen-react-components/es/MultiSelectSideBySide/items/item.js
import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import { Checkbox } from "@blueprintjs/core";
import ItemLabel from "./item_label";

import "./item.css";

var Item = function Item(_ref) {
  var _classnames;

  var item = _ref.item,
      height = _ref.height,
      _onClick = _ref.onClick,
      withBorder = _ref.withBorder,
      checked = _ref.checked,
      indeterminate = _ref.indeterminate,
      disabled = _ref.disabled;
  return (
    /* eslint-disable no-useless-computed-key*/

    React.createElement(
      "div",
      {
        className: classnames("mss-item", (_classnames = {}, _classnames["mss-with_border"] = withBorder, _classnames["mss-selected"] = checked, _classnames["mss-disabled"] = disabled, _classnames)),
        style: { height: height },
        onClick: function onClick(e) {
          if (_onClick) {
            e.stopPropagation();
            e.preventDefault();
            _onClick(e);
          }
        }
        /* eslint-enable no-useless-computed-key*/
      },
      React.createElement(Checkbox, {
        style: {
          marginTop: "10px",
          marginLeft: "10px"
        },
        type: "checkbox",
        color: "primary",
        checked: checked,
        indeterminate: indeterminate,
        disabled: disabled
      }),
      React.createElement(ItemLabel, { label: item.label })
    )
  );
};

Item.propTypes = process.env.NODE_ENV !== "production" ? {
  item: PropTypes.object,
  height: PropTypes.number,
  withBorder: PropTypes.bool,
  checked: PropTypes.bool,
  indeterminate: PropTypes.bool,
  disabled: PropTypes.bool
} : {};

Item.defaultProps = {
  item: {},
  height: 40,
  withBorder: false,
  checked: false,
  indeterminate: false,
  disabled: false
};

export default Item;
