// teselagen-react-components/lib/MultiSelectSideBySide/search/search.js
"use strict";

exports.__esModule = true;
exports.SearchWithValue = undefined;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _core = require("@blueprintjs/core");

require("./search.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Search = function Search(_ref) {
  var searchPlaceholder = _ref.searchPlaceholder,
      searchIcon = _ref.searchIcon,
      onChange = _ref.onChange;

  return _react2.default.createElement(
    "div",
    { className: "mss-search" },
    _react2.default.createElement("input", {
      type: "text",
      className: "mss-input",
      placeholder: searchPlaceholder,
      onChange: onChange
    }),
    _react2.default.createElement(
      "div",
      { className: "mss-icon" },
      _react2.default.createElement(_core.Icon, { icon: searchIcon })
    )
  );
};

var SearchWithValue = exports.SearchWithValue = function SearchWithValue(_ref2) {
  var searchPlaceholder = _ref2.searchPlaceholder,
      onChange = _ref2.onChange,
      value = _ref2.value,
      searchIcon = _ref2.searchIcon;

  return _react2.default.createElement(
    "div",
    { className: "mss-search" },
    _react2.default.createElement("input", {
      value: value,
      type: "text",
      className: "mss-input",
      placeholder: searchPlaceholder,
      onChange: onChange
    }),
    _react2.default.createElement(
      "div",
      { className: "mss-icon" },
      _react2.default.createElement(_core.Icon, { icon: searchIcon })
    )
  );
};

Search.propTypes = process.env.NODE_ENV !== "production" ? {
  searchPlaceholder: _propTypes2.default.string,
  searchIcon: _propTypes2.default.string,
  onChange: _propTypes2.default.func
} : {};

Search.defaultProps = {
  searchPlaceholder: "Search...",
  searchIcon: "search"
};

exports.default = Search;
